integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java

```diff
@@ -0,0 +1,281 @@
+      applicationMaster.start();
+      applicationMaster.requestContainers();
+      applicationMaster.stop();
+    } catch (Exception ex) {
+      LOG.error("Error running Application Master " + ex);
+      System.exit(1);
+    }
+  }
+
+  @Override
+  public void onContainersAllocated(List<Container> containers) {
+    if (!mMasterContainerAllocated) {
+      launchTachyonMasterContainers(containers);
+    } else {
+      launchTachyonWorkerContainers(containers);
+    }
+  }
+
+  @Override
+  public void onContainersCompleted(List<ContainerStatus> statuses) {
+    for (ContainerStatus status : statuses) {
+      LOG.error("Completed container " + status.getContainerId());
+    }
+  }
+
+  @Override
+  public void onNodesUpdated(List<NodeReport> updated) {}
+
+  @Override
+  public void onShutdownRequest() {
+    mApplicationDone = true;
+  }
+
+  @Override
+  public void onError(Throwable t) {}
+
+  @Override
+  public float getProgress() {
+    return 0;
+  }
+
+  public void start() throws IOException, YarnException {
+    // create a client to talk to NodeManager
+    mNMClient = NMClient.createNMClient();
+    mNMClient.init(mYarnConf);
+    mNMClient.start();
+
+    // Create a client to talk to the ResourceManager
+    mRMClient = AMRMClientAsync.createAMRMClientAsync(100, this);
+    mRMClient.init(mYarnConf);
+    mRMClient.start();
+
+    // Register with ResourceManager
+    mRMClient.registerApplicationMaster("" /* hostname */, 0 /* port */, "" /* tracking url */);
+    LOG.info("ApplicationMaster registered");
+  }
+
+  public void requestContainers() throws Exception {
+    // Priority for Tachyon master and worker containers - priorities are intra-application
+    Priority priority = Records.newRecord(Priority.class);
+    priority.setPriority(0);
+
+    // Resource requirements for master containers
+    Resource masterResource = Records.newRecord(Resource.class);
+    masterResource.setMemory(mMasterMemInMB);
+    masterResource.setVirtualCores(mMasterCpu);
+
+    String[] nodes = {mMasterAddress};
+
+    // Make container request for Tachyon master to ResourceManager
+    ContainerRequest masterContainerAsk =
+        new ContainerRequest(masterResource, nodes, null /* any racks */, priority);
+    LOG.info("Making resource request for Tachyon master on node " + mMasterAddress);
+    mRMClient.addContainerRequest(masterContainerAsk);
+
+    // Wait until Tachyon master container has been allocated
+    while (!mMasterContainerAllocated) {
+      Thread.sleep(1000);
+    }
+
+    // Resource requirements for master containers
+    Resource workerResource = Records.newRecord(Resource.class);
+    workerResource.setMemory(mWorkerMemInMB);
+    workerResource.setVirtualCores(mWorkerCpu);
+
+    // Make container requests for workers to ResourceManager
+    for (int i = 0; i < mNumWorkers; i ++) {
+      ContainerRequest containerAsk = new ContainerRequest(workerResource, null /* any hosts */,
```

Comment:
One quick question about this line. since location hint is ANY.  how can we ensure there won't be two workers on the same physical node?